File: 2021/03/advent202103.js
const fs = require('fs');
const path = require('path');

let input = fs.readFileSync(path.resolve(__dirname, 'input202103.txt'), 'utf-8', (err, data) => {
  if (err) {
    console.log(err)
  } else {
    return data;
  }
});

input = input.split('\n');

const bits = new Array(input[0].length).fill(0);

for (const line of input) {
  for (let i = 0; i < line.length; i++) {
    if (line[i] === '1') {
      bits[i]++;
    }
  }
}

let gamma = '';
let epsilon = '';
for (const bit of bits) {
  if (bit > input.length / 2) {
    gamma += '1';
    epsilon += '0';
  } else {
    gamma += '0';
    epsilon += '1';
  }
}
gamma = Number.parseInt(gamma, 2);
epsilon = Number.parseInt(epsilon, 2);
console.log(gamma * epsilon);

let place = 0;
let candidates = input;
while (candidates.length > 1) {
  let bit = 0;
  const ones = [];
  const zeros = [];
  for (const line of candidates) {
    if (line[place] === '1') {
      bit++;
      ones.push(line);
    } else {
      zeros.push(line);
    }
  }
  if (bit >= candidates.length / 2) {
    candidates = ones;
  } else {
    candidates = zeros;
  }
  place++;
}

const oxygenRating = Number.parseInt(candidates[0], 2);

place = 0;
candidates = input;
while (candidates.length > 1) {
  let bit = 0;
  const ones = [];
  const zeros = [];
  for (const line of candidates) {
    if (line[place] === '1') {
      bit++;
      ones.push(line);
    } else {
      zeros.push(line);
    }
  }
  if (bit >= candidates.length / 2) {
    candidates = zeros;
  } else {
    candidates = ones;
  }
  place++;
}

const CO2Rating = Number.parseInt(candidates[0], 2);

console.log(oxygenRating * CO2Rating);

File: 2021/03/advent202103.py
from typing import List

def find_common_bit(lines: List[str], place: int) -> int:
  bit = 0
  for line in lines:
    bit += 1 if line[place] == '1' else 0
  return 1 if bit >= len(lines) - bit else 0

def part1(lines: List[str]) -> int:
  gamma: List[str] = []
  epsilon: List[str] = []
  for i in range(len(lines[0])):
    bit = find_common_bit(lines, i)
    gamma.append(str(bit))
    epsilon.append(str(1 - bit))
  return int(''.join(gamma), 2) * int(''.join(epsilon), 2)

def part2(lines: List[str]) -> int:
  oxygen = lines
  co2 = lines
  place = 0
  while len(oxygen) > 1:
    bit = find_common_bit(oxygen, place)
    oxygen = [line for line in oxygen if line[place] == str(bit)]
    place += 1
  place = 0
  while len(co2) > 1:
    bit = 1 - find_common_bit(co2, place)
    co2 = [line for line in co2 if line[place] == str(bit)]
    place += 1
  return int(''.join(oxygen.pop()), 2) * int(''.join(co2.pop()), 2)

if __name__ == "__main__":
  with open('input.txt','r') as file:
    lines = [(line.strip()) for line in file]
  print(part1(lines))
  print(part2(lines))
